File: backend/backup/feature-flag-service/go/internal/model/flag_rule.go
package model

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// FlagRule represents a rule for evaluating a feature flag
type FlagRule struct {
	ID          string     `json:"id"`
	Name        string     `json:"name,omitempty"`
	Description string     `json:"description,omitempty"`
	Priority    int        `json:"priority"`
	Conditions  []Condition `json:"conditions"`
	Value       *FlagValue `json:"value"`
	Variations  []Variation `json:"variations,omitempty"`
}

// Condition represents a condition for a flag rule
type Condition struct {
	Attribute string      `json:"attribute"`
	Operator  string      `json:"operator"`
	Value     interface{} `json:"value"`
}

// Variation represents a variation of a flag value
type Variation struct {
	ID          string     `json:"id"`
	Value       *FlagValue `json:"value"`
	Weight      int        `json:"weight"`
	Description string     `json:"description,omitempty"`
}

// Operators
const (
	OperatorEquals              = "equals"
	OperatorNotEquals           = "notEquals"
	OperatorGreaterThan         = "greaterThan"
	OperatorGreaterThanOrEquals = "greaterThanOrEquals"
	OperatorLessThan            = "lessThan"
	OperatorLessThanOrEquals    = "lessThanOrEquals"
	OperatorContains            = "contains"
	OperatorNotContains         = "notContains"
	OperatorStartsWith          = "startsWith"
	OperatorEndsWith            = "endsWith"
	OperatorMatches             = "matches"
	OperatorIn                  = "in"
	OperatorNotIn               = "notIn"
)

// NewFlagRule creates a new flag rule
func NewFlagRule(id, name string, priority int, value *FlagValue) *FlagRule {
	return &FlagRule{
		ID:         id,
		Name:       name,
		Priority:   priority,
		Value:      value,
		Conditions: make([]Condition, 0),
		Variations: make([]Variation, 0),
	}
}

// AddCondition adds a condition to the rule
func (r *FlagRule) AddCondition(attribute, operator string, value interface{}) {
	r.Conditions = append(r.Conditions, Condition{
		Attribute: attribute,
		Operator:  operator,
		Value:     value,
	})
}

// AddVariation adds a variation to the rule
func (r *FlagRule) AddVariation(id string, value *FlagValue, weight int) {
	r.Variations = append(r.Variations, Variation{
		ID:     id,
		Value:  value,
		Weight: weight,
	})
}

// Evaluate evaluates the rule against the given context
func (r *FlagRule) Evaluate(context map[string]interface{}) (bool, error) {
	// If there are no conditions, the rule matches
	if len(r.Conditions) == 0 {
		return true, nil
	}

	// All conditions must match
	for _, condition := range r.Conditions {
		match, err := evaluateCondition(condition, context)
		if err != nil {
			return false, err
		}
		if !match {
			return false, nil
		}
	}

	return true, nil
}

// evaluateCondition evaluates a single condition against the context
func evaluateCondition(condition Condition, context map[string]interface{}) (bool, error) {
	// Get the attribute value from the context
	attrValue, exists := getAttributeValue(condition.Attribute, context)
	if !exists {
		// If the attribute doesn't exist, the condition doesn't match
		return false, nil
	}

	// Evaluate the condition based on the operator
	switch condition.Operator {
	case OperatorEquals:
		return equals(attrValue, condition.Value)
	case OperatorNotEquals:
		eq, err := equals(attrValue, condition.Value)
		return !eq, err
	case OperatorGreaterThan:
		return greaterThan(attrValue, condition.Value)
	case OperatorGreaterThanOrEquals:
		return greaterThanOrEquals(attrValue, condition.Value)
	case OperatorLessThan:
		return lessThan(attrValue, condition.Value)
	case OperatorLessThanOrEquals:
		return lessThanOrEquals(attrValue, condition.Value)
	case OperatorContains:
		return contains(attrValue, condition.Value)
	case OperatorNotContains:
		c, err := contains(attrValue, condition.Value)
		return !c, err
	case OperatorStartsWith:
		return startsWith(attrValue, condition.Value)
	case OperatorEndsWith:
		return endsWith(attrValue, condition.Value)
	case OperatorMatches:
		return matches(attrValue, condition.Value)
	case OperatorIn:
		return in(attrValue, condition.Value)
	case OperatorNotIn:
		i, err := in(attrValue, condition.Value)
		return !i, err
	default:
		return false, fmt.Errorf("unknown operator: %s", condition.Operator)
	}
}

// getAttributeValue gets the value of an attribute from the context
// Supports nested attributes with dot notation (e.g., "user.id")
func getAttributeValue(attribute string, context map[string]interface{}) (interface{}, bool) {
	parts := strings.Split(attribute, ".")
	current := context

	// Navigate through nested objects
	for i, part := range parts {
		if i == len(parts)-1 {
			// Last part, get the value
			val, exists := current[part]
			return val, exists
		}

		// Not the last part, navigate to the next level
		next, exists := current[part]
		if !exists {
			return nil, false
		}

		// Check if the next level is a map
		nextMap, ok := next.(map[string]interface{})
		if !ok {
			return nil, false
		}

		current = nextMap
	}

	return nil, false
}

// equals checks if two values are equal
func equals(a, b interface{}) (bool, error) {
	// Convert to comparable types
	aStr := fmt.Sprintf("%v", a)
	bStr := fmt.Sprintf("%v", b)
	return aStr == bStr, nil
}

// greaterThan checks if a > b
func greaterThan(a, b interface{}) (bool, error) {
	// Try to convert to numbers
	aFloat, aErr := toFloat64(a)
	bFloat, bErr := toFloat64(b)

	if aErr == nil && bErr == nil {
		// Both are numbers
		return aFloat > bFloat, nil
	}

	// Compare as strings
	aStr := fmt.Sprintf("%v", a)
	bStr := fmt.Sprintf("%v", b)
	return aStr > bStr, nil
}

// greaterThanOrEquals checks if a >= b
func greaterThanOrEquals(a, b interface{}) (bool, error) {
	// Try to convert to numbers
	aFloat, aErr := toFloat64(a)
	bFloat, bErr := toFloat64(b)

	if aErr == nil && bErr == nil {
		// Both are numbers
		return aFloat >= bFloat, nil
	}

	// Compare as strings
	aStr := fmt.Sprintf("%v", a)
	bStr := fmt.Sprintf("%v", b)
	return aStr >= bStr, nil
}

// lessThan checks if a < b
func lessThan(a, b interface{}) (bool, error) {
	// Try to convert to numbers
	aFloat, aErr := toFloat64(a)
	bFloat, bErr := toFloat64(b)

	if aErr == nil && bErr == nil {
		// Both are numbers
		return aFloat < bFloat, nil
	}

	// Compare as strings
	aStr := fmt.Sprintf("%v", a)
	bStr := fmt.Sprintf("%v", b)
	return aStr < bStr, nil
}

// lessThanOrEquals checks if a <= b
func lessThanOrEquals(a, b interface{}) (bool, error) {
	// Try to convert to numbers
	aFloat, aErr := toFloat64(a)
	bFloat, bErr := toFloat64(b)

	if aErr == nil && bErr == nil {
		// Both are numbers
		return aFloat <= bFloat, nil
	}

	// Compare as strings
	aStr := fmt.Sprintf("%v", a)
	bStr := fmt.Sprintf("%v", b)
	return aStr <= bStr, nil
}

// contains checks if a contains b
func contains(a, b interface{}) (bool, error) {
	aStr := fmt.Sprintf("%v", a)
	bStr := fmt.Sprintf("%v", b)
	return strings.Contains(aStr, bStr), nil
}

// startsWith checks if a starts with b
func startsWith(a, b interface{}) (bool, error) {
	aStr := fmt.Sprintf("%v", a)
	bStr := fmt.Sprintf("%v", b)
	return strings.HasPrefix(aStr, bStr), nil
}

// endsWith checks if a ends with b
func endsWith(a, b interface{}) (bool, error) {
	aStr := fmt.Sprintf("%v", a)
	bStr := fmt.Sprintf("%v", b)
	return strings.HasSuffix(aStr, bStr), nil
}

// matches checks if a matches the regex pattern b
func matches(a, b interface{}) (bool, error) {
	aStr := fmt.Sprintf("%v", a)
	bStr := fmt.Sprintf("%v", b)

	// Compile the regex pattern
	pattern, err := regexp.Compile(bStr)
	if err != nil {
		return false, fmt.Errorf("invalid regex pattern: %v", err)
	}

	return pattern.MatchString(aStr), nil
}

// in checks if a is in the list b
func in(a, b interface{}) (bool, error) {
	// Convert a to string for comparison
	aStr := fmt.Sprintf("%v", a)

	// Check if b is a slice
	switch bVal := b.(type) {
	case []interface{}:
		for _, item := range bVal {
			itemStr := fmt.Sprintf("%v", item)
			if aStr == itemStr {
				return true, nil
			}
		}
		return false, nil
	case string:
		// Split the string by comma
		items := strings.Split(bVal, ",")
		for _, item := range items {
			if aStr == strings.TrimSpace(item) {
				return true, nil
			}
		}
		return false, nil
	default:
		return false, fmt.Errorf("expected a list for 'in' operator, got %T", b)
	}
}

// toFloat64 converts a value to float64
func toFloat64(val interface{}) (float64, error) {
	switch v := val.(type) {
	case float64:
		return v, nil
	case float32:
		return float64(v), nil
	case int:
		return float64(v), nil
	case int64:
		return float64(v), nil
	case int32:
		return float64(v), nil
	case string:
		return strconv.ParseFloat(v, 64)
	default:
		return 0, fmt.Errorf("cannot convert %T to float64", val)
	}
}

File: backend/backup/feature-flag-service/python/feature_flag_service/model/flag_rule.py
"""
Flag rule module for the Feature Flag Service.
"""

import re
from typing import Any, Dict, List, Optional, Union

from pydantic import BaseModel, Field

from feature_flag_service.model.flag_value import FlagValue


class Condition(BaseModel):
    """
    Model for a condition in a flag rule.
    """
    attribute: str
    operator: str
    value: Any


class Variation(BaseModel):
    """
    Model for a variation in a flag rule.
    """
    id: str
    value: FlagValue
    weight: int
    description: Optional[str] = None


class FlagRule(BaseModel):
    """
    Model for a flag rule.
    """
    id: str
    name: Optional[str] = None
    description: Optional[str] = None
    priority: int
    conditions: List[Condition] = Field(default_factory=list)
    value: FlagValue
    variations: List[Variation] = Field(default_factory=list)

    # Operators
    OPERATOR_EQUALS = "equals"
    OPERATOR_NOT_EQUALS = "notEquals"
    OPERATOR_GREATER_THAN = "greaterThan"
    OPERATOR_GREATER_THAN_OR_EQUALS = "greaterThanOrEquals"
    OPERATOR_LESS_THAN = "lessThan"
    OPERATOR_LESS_THAN_OR_EQUALS = "lessThanOrEquals"
    OPERATOR_CONTAINS = "contains"
    OPERATOR_NOT_CONTAINS = "notContains"
    OPERATOR_STARTS_WITH = "startsWith"
    OPERATOR_ENDS_WITH = "endsWith"
    OPERATOR_MATCHES = "matches"
    OPERATOR_IN = "in"
    OPERATOR_NOT_IN = "notIn"

    def add_condition(self, attribute: str, operator: str, value: Any) -> None:
        """
        Add a condition to the rule.
        """
        self.conditions.append(Condition(
            attribute=attribute,
            operator=operator,
            value=value,
        ))

    def add_variation(self, id: str, value: FlagValue, weight: int, description: Optional[str] = None) -> None:
        """
        Add a variation to the rule.
        """
        self.variations.append(Variation(
            id=id,
            value=value,
            weight=weight,
            description=description,
        ))

    def evaluate(self, context: Dict[str, Any]) -> bool:
        """
        Evaluate the rule against the given context.
        """
        # If there are no conditions, the rule matches
        if not self.conditions:
            return True

        # All conditions must match
        for condition in self.conditions:
            if not self._evaluate_condition(condition, context):
                return False

        return True

    def _evaluate_condition(self, condition: Condition, context: Dict[str, Any]) -> bool:
        """
        Evaluate a single condition against the context.
        """
        # Get the attribute value from the context
        attr_value = self._get_attribute_value(condition.attribute, context)
        if attr_value is None:
            # If the attribute doesn't exist, the condition doesn't match
            return False

        # Evaluate the condition based on the operator
        if condition.operator == self.OPERATOR_EQUALS:
            return self._equals(attr_value, condition.value)
        elif condition.operator == self.OPERATOR_NOT_EQUALS:
            return not self._equals(attr_value, condition.value)
        elif condition.operator == self.OPERATOR_GREATER_THAN:
            return self._greater_than(attr_value, condition.value)
        elif condition.operator == self.OPERATOR_GREATER_THAN_OR_EQUALS:
            return self._greater_than_or_equals(attr_value, condition.value)
        elif condition.operator == self.OPERATOR_LESS_THAN:
            return self._less_than(attr_value, condition.value)
        elif condition.operator == self.OPERATOR_LESS_THAN_OR_EQUALS:
            return self._less_than_or_equals(attr_value, condition.value)
        elif condition.operator == self.OPERATOR_CONTAINS:
            return self._contains(attr_value, condition.value)
        elif condition.operator == self.OPERATOR_NOT_CONTAINS:
            return not self._contains(attr_value, condition.value)
        elif condition.operator == self.OPERATOR_STARTS_WITH:
            return self._starts_with(attr_value, condition.value)
        elif condition.operator == self.OPERATOR_ENDS_WITH:
            return self._ends_with(attr_value, condition.value)
        elif condition.operator == self.OPERATOR_MATCHES:
            return self._matches(attr_value, condition.value)
        elif condition.operator == self.OPERATOR_IN:
            return self._in(attr_value, condition.value)
        elif condition.operator == self.OPERATOR_NOT_IN:
            return not self._in(attr_value, condition.value)
        else:
            # Unknown operator
            return False

    def _get_attribute_value(self, attribute: str, context: Dict[str, Any]) -> Optional[Any]:
        """
        Get the value of an attribute from the context.
        Supports nested attributes with dot notation (e.g., "user.id").
        """
        parts = attribute.split(".")
        current = context

        # Navigate through nested objects
        for i, part in enumerate(parts):
            if i == len(parts) - 1:
                # Last part, get the value
                return current.get(part)

            # Not the last part, navigate to the next level
            if part not in current or not isinstance(current[part], dict):
                return None

            current = current[part]

        return None

    def _equals(self, a: Any, b: Any) -> bool:
        """
        Check if two values are equal.
        """
        return str(a) == str(b)

    def _greater_than(self, a: Any, b: Any) -> bool:
        """
        Check if a > b.
        """
        try:
            return float(a) > float(b)
        except (ValueError, TypeError):
            return str(a) > str(b)

    def _greater_than_or_equals(self, a: Any, b: Any) -> bool:
        """
        Check if a >= b.
        """
        try:
            return float(a) >= float(b)
        except (ValueError, TypeError):
            return str(a) >= str(b)

    def _less_than(self, a: Any, b: Any) -> bool:
        """
        Check if a < b.
        """
        try:
            return float(a) < float(b)
        except (ValueError, TypeError):
            return str(a) < str(b)

    def _less_than_or_equals(self, a: Any, b: Any) -> bool:
        """
        Check if a <= b.
        """
        try:
            return float(a) <= float(b)
        except (ValueError, TypeError):
            return str(a) <= str(b)

    def _contains(self, a: Any, b: Any) -> bool:
        """
        Check if a contains b.
        """
        return str(b) in str(a)

    def _starts_with(self, a: Any, b: Any) -> bool:
        """
        Check if a starts with b.
        """
        return str(a).startswith(str(b))

    def _ends_with(self, a: Any, b: Any) -> bool:
        """
        Check if a ends with b.
        """
        return str(a).endswith(str(b))

    def _matches(self, a: Any, b: Any) -> bool:
        """
        Check if a matches the regex pattern b.
        """
        try:
            pattern = re.compile(str(b))
            return bool(pattern.match(str(a)))
        except re.error:
            return False

    def _in(self, a: Any, b: Any) -> bool:
        """
        Check if a is in the list b.
        """
        if isinstance(b, list):
            return str(a) in [str(item) for item in b]
        elif isinstance(b, str):
            items = [item.strip() for item in b.split(",")]
            return str(a) in items
        return False

